pipelines/salesforce/helpers.py

```diff
@@ -0,0 +1,68 @@
+import pendulum
+
+from typing import Optional, Iterable
+
+from simple_salesforce import Salesforce
+from dlt.common.typing import TDataItem
+
+
+from .settings import IS_PRODUCTION
+
+
+def get_records(
+    sf: Salesforce,
+    sobject: str,
+    last_state: Optional[str] = None,
+    replication_key: Optional[str] = None,
+) -> Iterable[TDataItem]:
+    """
+    Retrieves records from Salesforce for a specified sObject.
+
+    Args:
+        sf (Salesforce): An instance of the Salesforce API client.
+        sobject (str): The name of the sObject to retrieve records from.
+        last_state (str, optional): The last known state for incremental loading. Defaults to None.
+        replication_key (str, optional): The replication key for incremental loading. Defaults to None.
+
+    Yields:
+        Dict[str, Any]: A dictionary representing a record from the Salesforce sObject.
+    """
+
+    # Get all fields for the sobject
+    desc = getattr(sf, sobject).describe()
+    # Salesforce returns compound fields as separate fields, so we need to filter them out
+    compound_fields = {
+        f["compoundFieldName"]
+        for f in desc["fields"]
+        if f["compoundFieldName"] is not None
+    } - {"Name"}
+    # Salesforce returns datetime fields as timestamps, so we need to convert them
+    date_fields = {
+        f["name"] for f in desc["fields"] if f["type"] in ("datetime",) and f["name"]
+    }
+    # If no fields are specified, use all fields except compound fields
+    fields = [f["name"] for f in desc["fields"] if f["name"] not in compound_fields]
+
+    # Generate a predicate to filter records by the replication key
+    predicate, order_by, n_records = "", "", 0
+    if replication_key:
+        if last_state:
+            predicate = f"WHERE {replication_key} > {last_state}"
+        order_by = f"ORDER BY {replication_key} ASC"
+    query = f"SELECT {', '.join(fields)} FROM {sobject} {predicate} {order_by}"
```

Comment:
Outside of the scope of this PR: an improvement is to escape the values with [`format_soql`](https://github.com/simple-salesforce/simple-salesforce#queries):

```python
from simple_salesforce import format_soql
 
sf.query(format_soql("SELECT Id, Email FROM Contact WHERE LastName = {}", "Jones"))
sf.query(format_soql("SELECT Id, Email FROM Contact WHERE LastName = {last_name}", last_name="Jones"))
sf.query(format_soql("SELECT Id, Email FROM Contact WHERE LastName IN {names}", names=["Smith", "Jones"]))
```

Just in case 😉